### PR TITLE
NFC: Moved device related code out of TFUtilities into a separate header file

### PR DESCRIPTION
**Pull request by mhong**
Also addressed the following feedback from @lattner:

- Changed GraphGlobalConfiguration::usedDeviceTypes to be an array of bools
  indexed by the enum value. This then allowed us to remove the
  DenseMapInfo<::swift::tf::DeviceType> specialization.

This changes the ordering of iterating over the used device types. e.g. now we are guaranteed to generate a CPU device-partitioned graph function before GPU, because of the DeviceType enum ordering, regardless of which one is the "primary device type". As such, we could be hitting a "missing shapes attr error" when generating a TPU infeed enqueue op (on CPU), before we generate the corresponding TPU infeed dequeue op (on TPU).

- Renamed GraphGlobalConfiguration to GraphFunctionDeviceInfo.

- Simplified the handleDevicePlacement() API, and verified that when a second
  device attribute to added to the same graph_op inst, SILVerifier with assert
  fail.

- Changed DevicePartitionerImpl::instByDevice to be a C array:
  SmallPtrSet<SILInstruction *, 8> instByDevice[NUM_DEVICE_TYPES];

- Changed tfc.scalarToTensor to allow device placement, but it gets
  ignored (conceptually the promoted scalar is on ALL devices). This removes
  some special case code in the compiler.

### CODE REVIEW

**Comment by mhong** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 1]
```diff
@@ -13,14 +13,15 @@
...
+#include "TFDeviceSupport.h"
 #include "swift/AST/ASTContext.h"
 #include "swift/AST/DiagnosticsSIL.h"
 #include "swift/AST/GenericSignatureBuilder.h"
 #include "swift/AST/Module.h"
 #include "swift/SIL/SILBuilder.h"
 #include "swift/SIL/SILConstants.h"
 #include "swift/SIL/SILModule.h"
-#include "swift/SILOptimizer/Utils/Local.h"
+// #include "swift/SILOptimizer/Utils/Local.h"
```
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: comment can be dropped.

Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: If other #includes can be dropped, please do!

**Review by lattner — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 1 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Done. removed `#include "llvm/Support/FileSystem.h"` too!

File: lib/SILOptimizer/Mandatory/TFDeviceSupport.h [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+#ifndef SWIFT_SILOPTIMIZER_TFDEVICESUPPORT_H
+#define SWIFT_SILOPTIMIZER_TFDEVICESUPPORT_H
+
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILConstants.h"
+#include "swift/SIL/SILLocation.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/ErrorHandling.h"
```
[the remark this answers is not in the document]
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFDeviceSupport.h]: This has been pruned.

File: lib/SILOptimizer/Mandatory/TFDeviceSupport.h [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This defines the shared code that implements the various TensorFlow related
+// lowerings and other transformations.
```
[the remark this answers is not in the document]
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFDeviceSupport.h]: Done

File: lib/SILOptimizer/Mandatory/TFDeviceSupport.h [hunk 4]
```diff
@@ -0,0 +1,324 @@
+//===--- TFUtilities.h - TensorFlow lowering utilities ----------*- C++ -*-===//
```
[the remark this answers is not in the document]
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFDeviceSupport.h]: Done

**Comment by mhong** [comment]
Thanks @lattner for the quick review! I'm merging this PR as it touches many files and chunks, to minimize potential conflicts. @rxwei, I'd be happy to address any additional comments.

**Comment by rxwei** [comment]
LGTM! Would've been easier to review if the code move was separate from code changes. GitHub diff sucks :(

**Comment by mhong** [comment]
Agree. This turned out to be a larger change than I expected. I should factor this into separate PRs for future scenarios.

**Pull request merged**